### PR TITLE
[runtime] Enhancement: add new demi_wait_next_n runtime binding

### PR DESCRIPTION
**Pull request by kyleholohan**
This PR adds a new binding to get the next (n) completed operation(s). I also added some annotations to the bindings to engage some correctness checking on the behalf of GCC/clang/MSVC.

### CODE REVIEW

File: include/demi/types.h [hunk 1]
```diff
@@ -21,6 +21,35 @@ extern "C"
...
+#  define ATTR_NONNULL(...) __attribute__((nonnull(__VA_ARGS__)))
+#  define ATTR_NODISCARD __attribute__((warn_unused_result))
+#elif defined(_MSC_VER)
+// MSVC uses SAL; NONNULL is supported via _In_/_Out_/etc.
+#  define ATTR_NONNULL(...)
+#  define ATTR_NODISCARD _Check_return_
+#else
+#  define ATTR_NONNULL(...)
+#  define ATTR_NODISCARD _Check_return_
+#endif
```
Comment (ppenna) [inline include/demi/types.h:51]: I would suggest putting this on `<demi/cc.h>` as it concerns compiler-glue code.

**Review by ppenna — APPROVED** [review]
LGTM

**Review by iyzhang — APPROVED** [review]
LGTM

**Comment by iyzhang** [comment]
This PR closes #1134

**Pull request merged**